### PR TITLE
Add token auction system contract bytecode

### PR DESCRIPTION
**Pull request by hanyunx**

### CODE REVIEW

File: quarkchain/cluster/tests/test_shard_state.py [hunk 1]
```diff
@@ -3158,3 +3158,96 @@ def test_remove_tx_from_queue_with_higher_nonce(self):
...
         self.assertEqual(len(state.tx_queue), 0)
+
+    def test_mint_new_native_token(self):
+        id1 = Identity.create_random_identity()
+        acc1 = Address.create_from_identity(id1, full_shard_key=0)
+        env = get_test_env(genesis_account=acc1, genesis_minor_quarkash=10 ** 20)
+        state = create_default_shard_state(env=env)
+        evm_state = state.evm_state  # type: EvmState
+
+        contract_runtime_bytecode = bytes.fromhex(
```
Comment (ninjaahhh) [inline quarkchain/cluster/tests/test_shard_state.py]: can refer to #792 on how to get runtime code from deployment bytecode. same for function invocation bytecode

File: quarkchain/cluster/tests/test_shard_state.py [hunk 2]
```diff
@@ -3158,3 +3158,96 @@ def test_remove_tx_from_queue_with_higher_nonce(self):
...
+        )
+        env.quark_chain_config.NON_RESERVED_NATIVE_TOKEN_CONTRACT_BYTECODE_HASH = sha3_256(
+            contract_runtime_bytecode
+        ).hex()
+        contract_addr = SystemContract.NON_RESERVED_NATIVE_TOKEN.addr()
+        evm_state.set_code(contract_addr, contract_runtime_bytecode)
+        evm_state.set_storage_data(contract_addr, 0, acc1.recipient)
+        evm_state.timestamp = 1
```
Comment (ninjaahhh) [inline quarkchain/cluster/tests/test_shard_state.py]: use `int(time.time())`?

File: quarkchain/cluster/tests/test_shard_state.py [hunk 3]
```diff
@@ -3158,3 +3158,96 @@ def test_remove_tx_from_queue_with_higher_nonce(self):
...
+                data=bytes.fromhex(data),
+            ).tx.to_evm_tx()
+            ret.set_quark_chain_config(env.quark_chain_config)
+            return ret
+
+        # set minPriceInQKC to be 20, minIncrement to be 5%, duration to be one week (604800 s)
+        set_auction_params = lambda n, v: tx_gen(
+            n,
+            v,
+            "3c69e3d2000000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000000050000000000000000000000000000000000000000000000000000000000093a80",
```
Comment (ninjaahhh) [inline quarkchain/cluster/tests/test_shard_state.py]: as mentioned, you can make this more readable

File: quarkchain/cluster/tests/test_shard_state.py [hunk 4]
```diff
@@ -3166,19 +3169,22 @@ def test_mint_new_native_token(self):
...
-        )
+        runtime_bytecode = NON_RESERVED_NATIVE_TOKEN_CONTRACT_BYTECODE
+        runtime_start = runtime_bytecode.find(bytes.fromhex("608060405260"), 1)
+        runtime_bytecode = runtime_bytecode[runtime_start:-32]
         env.quark_chain_config.NON_RESERVED_NATIVE_TOKEN_CONTRACT_BYTECODE_HASH = sha3_256(
```
Comment (ninjaahhh) [inline quarkchain/cluster/tests/test_shard_state.py]: I don't think you need this

File: quarkchain/cluster/tests/test_shard_state.py [hunk 5]
```diff
@@ -3190,64 +3196,59 @@ def tx_gen(nonce, value, data: str):
...
+        tx3 = end_auction()
         success, _ = apply_transaction(evm_state, tx3, bytes(32))
         self.assertTrue(success)
 
-        tx4 = get_native_token_info(4, 0)
+        tx4 = get_native_token_info()
         success, output = apply_transaction(evm_state, tx4, bytes(32))
         self.assertTrue(success)
-        self.assertEqual(int.from_bytes(output[:32], byteorder="big"), 604801)
+        self.assertNotEqual(int.from_bytes(output[:32], byteorder="big"), 0)
```
Comment (ninjaahhh) [inline quarkchain/cluster/tests/test_shard_state.py]: the value should equal to the owner, which is `acc1.recipient`?

Comment (hanyunx) [reply quarkchain/cluster/tests/test_shard_state.py]: It's the created time of queried token.

File: quarkchain/cluster/tests/test_shard_state.py [hunk 6]
```diff
@@ -3158,3 +3161,93 @@ def test_remove_tx_from_queue_with_higher_nonce(self):
...
+                data=bytes.fromhex(data),
+            ).tx.to_evm_tx()
+            nonce += 1
+            ret.set_quark_chain_config(env.quark_chain_config)
+            return ret
+
+        token_id = 9999999  # token id to bid and win
+        amount = 1000
+        parsed_hex = lambda i: i.to_bytes(32, byteorder="big").hex()
+        set_auction_params = lambda: tx_gen(
```
Comment (ninjaahhh) [inline quarkchain/cluster/tests/test_shard_state.py]: add a comment explaining what are the parameters

File: quarkchain/cluster/tests/test_shard_state.py [hunk 7]
```diff
@@ -3158,3 +3161,93 @@ def test_remove_tx_from_queue_with_higher_nonce(self):
...
+        tx1 = resume_auction()
+        success, _ = apply_transaction(evm_state, tx1, bytes(32))
+        self.assertTrue(success)
+
+        tx2 = bid_new_token(26 * 10 ** 18)
+        success, _ = apply_transaction(evm_state, tx2, bytes(32))
+        self.assertTrue(success)
+
+        # 7 days passed, this round of auction ends
+        evm_state.timestamp += 3600 * 24 * 7
```
Comment (ninjaahhh) [inline quarkchain/cluster/tests/test_shard_state.py]: add another case that ending auction before the ending time should fail

File: quarkchain/config.py [hunk 8]
```diff
@@ -320,6 +320,9 @@ class QuarkChainConfig(BaseConfig):
     ROOT_CHAIN_POSW_CONTRACT_BYTECODE_HASH = (
         "0000000000000000000000000000000000000000000000000000000000000000"
     )
+    NON_RESERVED_NATIVE_TOKEN_CONTRACT_BYTECODE_HASH = (
```
Comment (ninjaahhh) [inline quarkchain/config.py]: this is not being used, can delete

File: quarkchain/tests/test_config.py [hunk 9]
```diff
@@ -203,7 +203,8 @@ def test_serialization(self):
     "MIN_MINING_GAS_PRICE": 1000000000,
     "XSHARD_GAS_DDOS_FIX_ROOT_HEIGHT": 90000,
     "DISABLE_POW_CHECK": false,
-    "ROOT_CHAIN_POSW_CONTRACT_BYTECODE_HASH": "0000000000000000000000000000000000000000000000000000000000000000"
+    "ROOT_CHAIN_POSW_CONTRACT_BYTECODE_HASH": "0000000000000000000000000000000000000000000000000000000000000000",
+    "NON_RESERVED_NATIVE_TOKEN_CONTRACT_BYTECODE_HASH": "0000000000000000000000000000000000000000000000000000000000000000"
```
Comment (ninjaahhh) [inline quarkchain/tests/test_config.py]: same

**Pull request merged**